Add RooMultiPdf

fionnbishop opened the pull request:
This adds the possibility of using the discrete profiling method https://arxiv.org/abs/1408.6865 when setting limits on datasets. The user just needs to put the multipdf name in initPDF and the code will automatically detect that it is a RooMultiPdf and act accordingly.

Each fit iterates over the discrete nuisance parameter and returns the RooFitResult with the lowest NLL (after applying an ndof correction). However, if one of the fits fails, we want the whole set of fits to fail, so in this case, it breaks the loop and returns the failed RooFitResult (below). The failure criteria `result_tmp->status()!=0 or result_tmp->covQual()!=3 or result_tmp->edm()>1.e-3` do have to be specified manually, and should be the same as those in MethodDatasetsPluginScan.
https://github.com/gammacombo/gammacombo/blob/3508f61a1a5fe0a32d81f4f4a1f860915f90903f/core/src/PDF_Datasets.cpp#L354-L365



Toys are generated using the nuisance parameters which minimise the NLL at that scanpoint; I have implemented this behaviour for discrete nuisance parameters. I tried to keep changes to other parts of the code to a minimum.

The best fit nuisance parameters for the B-only fit are found here and this->pdf->bestIndexBkg is assigned the best index
https://github.com/gammacombo/gammacombo/blob/ecc2bf8753630b14bf5c083d6c8735bfd123367c/core/src/MethodDatasetsPluginScan.cpp#L125-L126
This best index is picked up by PDF_Datasets::fitBkg
https://github.com/gammacombo/gammacombo/blob/ecc2bf8753630b14bf5c083d6c8735bfd123367c/core/src/PDF_Datasets.cpp#L558-L560

The best fit nuisance parameters for the S+B fit are found in MethodDatasetsProbScan. It is necessary to store both the best discrete nuisance parameter and the values of the continuous nuisance parameters which correspond to that discrete parameter (as opposed to just the nuisance parameter values in the workspace, as before).
https://github.com/gammacombo/gammacombo/blob/ecc2bf8753630b14bf5c083d6c8735bfd123367c/core/src/MethodDatasetsProbScan.cpp#L464-L475

fionnbishop commented:
The only sticking point I can think of is that if a fit fails in the ProbScan, it will generate toys which are very different to at the other scan points, potentially causing discontinuities in the CLs curve. This is potentially a problem in general, but I think it is worse when using a multipdf as the pdf variation is large when the discrete nuisance parameter is different. It's not obvious what the best solution is.

tmombaecher commented:
Is this ready for review? As for your last point I agree this is not trivial. But for the "standard" scan the fit should (if ever) fail at the boundaries of the scan. If it's not possible to restrict the scan range to a robust one, one should reduce fit complexity.
As for the MultiPdf one could implement a refit excluding the index with the failing model. But this might create a small bias (probably not too problematical though).

tmombaecher commented on core/src/PDF_Datasets.cpp:
```diff
@@ -167,12 +196,24 @@ void PDF_Datasets::initBkgPDF(const TString& name) {
     }
     pdfBkgName  = name;
-    pdfBkg      = wspc->pdf(pdfBkgName);
+    if (wspc->pdf(pdfBkgName)->InheritsFrom("RooMultiPdf")) {
```
I would like to preserve the possibility to supply a background-only pdf (though often it's better not to such that it's defined by sig=0). This line here might lead to an unintelligble SegFault if the wspc doesn't contain the bkgpdf. So suggest to protect against this by first asking whether the component exists in the wspc before asking `InheritsFrom`

fionnbishop replied:
Thanks for pointing this out. I have dealt with this now, for initPDF as well.

tmombaecher replied:
Does this now mean that if you don't specify a bkg pdf, the scan will not work? I had in mind to provide the two functionalities for bkg-only fits:
1. if a bkg-only pdf is provided, use that one for fitting and generation of bkg toys
2. if no bkg fit is provided, use the signal+background pdf and set the parameter of interest to zero.
While option 1 is the more obvious method (and both methods should normally be equivalent), option 2 is often more stable for reasons I don't understand. Therefore I find it sensible to keep both options

fionnbishop replied:
It should still work - If using option 2, the initBkgPDF isn't called.
What I understand you want to happen is that if the user calls initPDF or initBkgPDF with a pdf name not in wspc, the code exits with an error message. As far as I can tell, this is what happens here.

fionnbishop commented:
> Is this ready for review? 

Almost - I thought it was but earlier today thought of a couple of things which need addressing. There is also your point above to fix. I'll hopefully be able to address these fairly quickly, tomorrow if not later today.

> As for your last point I agree this is not trivial. But for the "standard" scan the fit should (if ever) fail at the boundaries of the scan. If it's not possible to restrict the scan range to a robust one, one should reduce fit complexity.
> As for the MultiPdf one could implement a refit excluding the index with the failing model. But this might create a small bias (probably not too problematical though).
I would expect there could be a sizeable bias as a result of removing the higher complexity pdfs, so I wouldn't want to do this.
One partial solution could be to print out the fit status in ProbScan similarly to PluginScan (potentially just for failed fits)? This doesn't solve the problem, but it does at least show the user that there is a problem.

fionnbishop commented:
@tmombaecher I have now checked/dealt with all the things that I thought of.

tmombaecher reviewed (commented):
Thanks a lot for the work! I tested it so far and don't see major points. Just the question mentioned above.
Maybe I just pick up @matthewkenzie's comment that it would be great to add in the beginning of the new tutorial a few lines on how to use it (we could then even port that into the CI check) and show the effect the multipdf has with a plot or so

fionnbishop commented on core/src/PDF_Datasets.cpp:
```diff
@@ -301,17 +351,44 @@ RooFitResult* PDF_Datasets::fit(RooDataSet* dataToFit) {
+            RooAbsReal* nll = multipdf->getPdf(npdf)->createNLL(*dataToFit, RooFit::Extended(kTRUE), RooFit::ExternalConstraints(*getWorkspace()->set(constraintName)));
+            minMultipdfNll = nll->getVal()+multipdf->getCorrection();
+            if (result_tmp->status()!=0 or result_tmp->covQual()!=3 or result_tmp->edm()>1.e-3) badFit = true;
+            if (npdf==0 or minMultipdfNll < this->minNll or badFit) {
```
I suppose it doesn't really matter whether or not minNll is set if the fit is bad. The only important thing is that the bad fit result is saved so that subsequent parts of the code know that this is a failed fit. It is easier to set it since that is required for the other cases.

I'm not sure what this->minNll is initially, but I don't think it matters, since the statement is always triggered by npdf==0 on the first iteration.

tmombaecher replied:
ok, true

fionnbishop commented:
> 
> 
> Thanks a lot for the work! I tested it so far and don't see major points. Just the question mentioned above.
> Maybe I just pick up @matthewkenzie's comment that it would be great to add in the beginning of the new tutorial a few lines on how to use it (we could then even port that into the CI check) and show the effect the multipdf has with a plot or so

I have added more detail to the tutorial comments, so I think there should be enough information there for anyone to run it assuming they have already done the standard limits tutorial. (though it is hard to know what information needs to be included when working from the perspective of someone who already knows how it works!)

tmombaecher commented:
What I would find useful for the tutorial is one setup where you see the effect that the MultiPdf makes. I.e. one setup where you have only one model and one setup where you have two models and the MultiPdf takes care of incorporating the systematic. Is something like that possible?

fionnbishop commented:
> What I would find useful for the tutorial is one setup where you see the effect that the MultiPdf makes. I.e. one setup where you have only one model and one setup where you have two models and the MultiPdf takes care of incorporating the systematic. Is something like that possible?

This could be achieved by running the tutorial twice, once with initPDF("roomultipdf") and once with e.g. initPDF("SBmodel_exponential"). I can add a note about this, if that would be helpful.
I have tried to the effect of the multipdf somewhat clear, and the best pdf does change over the scanrange in this tutorial, however, it only changes at the high end, so doesn't have a substantial effect on the CLs. I have found that it is actually quite difficult to construct a dataset/set of pdfs that would produce a large effect.

tmombaecher commented:
Can you add the plots for the comparison of the suggestion in this thread? I think then we can merge.

fionnbishop commented:
OK, I first will have another fiddle with the dataset/pdf settings to try and make the effect as obvious as possible, and then I'll put those plots here.

fionnbishop commented:
[tutorial_dataset_branchingRatio_plugin_cls_MULTIPDF.pdf](https://github.com/gammacombo/gammacombo/files/5364070/tutorial_dataset_branchingRatio_plugin_cls_MULTIPDF.pdf)
[tutorial_dataset_branchingRatio_plugin_cls_POLYNOMIAL.pdf](https://github.com/gammacombo/gammacombo/files/5364069/tutorial_dataset_branchingRatio_plugin_cls_POLYNOMIAL.pdf)

@tmombaecher these are CLs/CL curves for the multipdf and for the pdf with a polynomial bkg model (I switched to using this as the comparison as it is the global minimum for this multipdf/dataset). It is not a large effect, but the points are definitely higher in the second half of the scan range when using the multipdf.
(I'm not sure why the prob curve is a bit spiky - But with a multipdf the prob curve is not completely meaningful, so I'm not too worried about that)

fionnbishop commented:
@tmombaecher On a separate note, a bug I spotted: Here a refit is done to find the global min if any variables in the workspace are outside of the physical range. However, the most recent fit seems to be a bkg-only fit, and the workspace reflects this. Therefore, the poi could be outside the physical range in the global fit, but inside at this point in the code.

https://github.com/gammacombo/gammacombo/blob/717a1c0d4723ef9cfff793a5cf0983db9837a461/core/src/MethodDatasetsPluginScan.cpp#L50-L75

I think this could be fixed in one line with Utils::setParameters(w,globalMin). Seeing as this is a one-line fix, I could add it to this pr, or seeing as it is unrelated, would it be better to make a new issue for this?

tmombaecher commented:
> @tmombaecher On a separate note, a bug I spotted: Here a refit is done to find the global min if any variables in the workspace are outside of the physical range. However, the most recent fit seems to be a bkg-only fit, and the workspace reflects this. Therefore, the poi could be outside the physical range in the global fit, but inside at this point in the code.
> 
> https://github.com/gammacombo/gammacombo/blob/717a1c0d4723ef9cfff793a5cf0983db9837a461/core/src/MethodDatasetsPluginScan.cpp#L50-L75
> 
> I think this could be fixed in one line with Utils::setParameters(w,globalMin). Seeing as this is a one-line fix, I could add it to this pr, or seeing as it is unrelated, would it be better to make a new issue for this?

Yes, this apparently slipped my attention when I went through and removed every unnecessary fit. Please just add this one-liner here, then I would merge

fionnbishop commented:
> Yes, this apparently slipped my attention when I went through and removed every unnecessary fit. Please just add this one-liner here, then I would merge
Fixed this, and also another bug in ProbScan where the parameters needed to be set to the fit result values in order to get the correct scanbest value when using a multipdf (don't think it affects the results, just the log output)
